clippy_lints/src/arithmetic.rs

```diff
@@ -70,30 +70,63 @@ impl<'tcx> LateLintPass<'tcx> for Arithmetic {
             }
         }
         match &expr.kind {
-            hir::ExprKind::Binary(op, l, r) | hir::ExprKind::AssignOp(op, l, r) => {
-                match op.node {
-                    hir::BinOpKind::And
-                    | hir::BinOpKind::Or
-                    | hir::BinOpKind::BitAnd
-                    | hir::BinOpKind::BitOr
-                    | hir::BinOpKind::BitXor
-                    | hir::BinOpKind::Eq
-                    | hir::BinOpKind::Lt
-                    | hir::BinOpKind::Le
-                    | hir::BinOpKind::Ne
-                    | hir::BinOpKind::Ge
-                    | hir::BinOpKind::Gt => return,
-                    _ => (),
-                }
+            hir::ExprKind::Binary(op, l, r) | hir::ExprKind::AssignOp(op, l, r) => match op.node {
+                hir::BinOpKind::And
+                | hir::BinOpKind::Or
+                | hir::BinOpKind::BitAnd
+                | hir::BinOpKind::BitOr
+                | hir::BinOpKind::BitXor
+                | hir::BinOpKind::Eq
+                | hir::BinOpKind::Lt
+                | hir::BinOpKind::Le
+                | hir::BinOpKind::Ne
+                | hir::BinOpKind::Ge
+                | hir::BinOpKind::Gt => (),
 
-                let (l_ty, r_ty) = (cx.typeck_results().expr_ty(l), cx.typeck_results().expr_ty(r));
-                if l_ty.peel_refs().is_integral() && r_ty.peel_refs().is_integral() {
-                    span_lint(cx, INTEGER_ARITHMETIC, expr.span, "integer arithmetic detected");
-                    self.expr_span = Some(expr.span);
-                } else if l_ty.peel_refs().is_floating_point() && r_ty.peel_refs().is_floating_point() {
-                    span_lint(cx, FLOAT_ARITHMETIC, expr.span, "floating-point arithmetic detected");
-                    self.expr_span = Some(expr.span);
-                }
+                hir::BinOpKind::Div | hir::BinOpKind::Rem => match &r.kind {
+                    hir::ExprKind::Lit(lit) => {
+                        if let rustc_ast::ast::LitKind::Int(0, _) = lit.node {
+                            span_lint(cx, INTEGER_ARITHMETIC, expr.span, "integer arithmetic detected");
+                            self.expr_span = Some(expr.span);
+                        }
+                        if let rustc_ast::ast::LitKind::Float(_, _) = lit.node {
```

Comment:
This code snippet appears multiple times. Could you combine all of them to a single occurrence (similar to what it was before)?
```
                        if let rustc_ast::ast::LitKind::Float(_, _) = lit.node {
                            span_lint(cx, FLOAT_ARITHMETIC, expr.span, "floating-point arithmetic detected");
                            self.expr_span = Some(expr.span);
                        }
```

Reply:
I tried to minimize the duplication, but even if I add a match statement there I would need to have two cases one for int and one for float because they have different messages. Also, I used `if let` twice rather than match statement to avoid another intendation.

Is there any other way to avoid duplicatoin that I am missing?

Reply:
I tried to minimize the duplication, but even if I add a match statement there I would need to have two cases one for int and one for float because they have different messages. Also, I used if let twice rather than match statement to avoid another indentation.

Is there any other way to avoid duplication that I am missing?

Reply:
You could keep the old code until line 89 and then do the case distinction for integer and float. Then you need to do another match for the integer case (to detect the division/remainder operation), but you could combine the four different occurrences of the floating-point lint.

Reply:
I managed to combine different occurrences of floating-point line. But I still had to repeat the integer lint four times.